utils/merge_and_push_translations.sh

```diff
@@ -54,89 +54,141 @@ fi
+  # - Authors
+  git add 'data/txts/*.lua' || true
+  # - Locale data
+  git add 'data/i18n/*.lua' || true
+  # - Appdata
+  git add xdg/org.widelands.Widelands.appdata.xml xdg/org.widelands.Widelands.desktop || true
+  # - Statistics
+  git add data/i18n/translation_stats.conf || true
+}
 
 undo_oneliner_diffs() {
   # Undo one-liner diffs of pure timestamps with no other content
   set +x
   for entry in $(git diff --numstat po/ | sed -En 's/^1\t1\t//p'); do
     if [ -z "$(git diff "$entry" | grep '^[+-][^+-]' | grep -v '^[+-]"POT-Creation-Date:')" ]
-    then
+    then # no other diff line remaining
       echo "Skipping changes to $entry"
-      git checkout $entry
+      git checkout "$entry"
     fi
   done
   set -x
 }
+
+# Print all commands.
+set -x
+
+# Pull All translations from Transifex
+# use force to make sure really all files get pulled
+tx pull -a -f
+
+# Undo one-liner diffs of pure timestamps with no other content
+# for fetched translations (*.po) (in case something plays tricks, see #5937)
+undo_oneliner_diffs
+
+run_update_scripts=at_least_once # run them at least once
+if [ -n "$(git status -s)" ]; then
+  update_authors
+  update_appdata
+  update_statistics
+  run_update_scripts= # they have run now
+
+  # Stage translations
+  git add 'po/*/*.po' 'data/i18n/locales/*.json' 'xdg/translations/*.json'
+  # and generated files
+  gitAddGeneratedFiles
+
+  # commit translations
+  git commit -m "Fetched translations and updated data."
+  had_commit=yes
+fi
+
+# -----------------------
+
+# Update source catalogs
+python3 utils/buildcat.py
+
+# Undo one-liner diffs of pure timestamps with no other content
+# for generated catalogs (*.pot)
 undo_oneliner_diffs
 
 if [ -n "$(git status -s)" ]; then
   # Only upload to Transifex if anything changed
   # Push source catalogs to Transifex
   tx push -s
   sleep 65 # wait for translation files to be updated
-fi
 
-# Pull All translations from Transifex
-# tx pull -a
-# Force-pull translations because some would get skipped accidentally
-tx pull -fa
-
-# Update authors file
-python3 utils/update_authors.py
-if [ $? -eq 0 ] ; then
-  echo "Updated authors";
-else
-  echo "Failed updating authors";
-  exit 1;
-fi
+  # Pull All translations from Transifex
+  # use force to make sure really all files get pulled
+  tx pull -a -f
 
-# Update appdata
-python3 utils/update_appdata.py
-if [ $? -eq 0 ] ; then
-  echo "Updated appdata";
-else
-  echo "Failed updating appdata";
-  exit 1;
-fi
+  # Undo one-liner diffs of pure timestamps with no other content
+  # for fetched translations (*.po) (in case something plays tricks, see #5937)
+  undo_oneliner_diffs
 
-# Update statistics
-python3 utils/update_translation_stats.py
-if [ $? -eq 0 ] ; then
-  echo "Updated translation stats";
-else
-  echo "Failed to update translation stats";
-  exit 1;
+  if [ -z "$run_update_scripts" ] && [ -n "$(git status -s)" ]; then
+    run_update_scripts=for_pot # run again, pot files have changed
+  fi
```

Comment:
```suggestion
  if [ -n "$(git status -s)" ]; then
    update_authors # in case something has changed unexpectedly
    update_appdata # in case something has changed unexpectedly
    update_statistics
  fi
```

So you can remove the `run_update_scripts` variable.

Reply:
The variable run_update_scripts is for case `d)` of [my comment](#issuecomment-1685073340). 
I want the update_* scripts to run once at least. As this has happened before this pr. So I run them before the second commit if they were not run for the 1st commit, even without change of a po file.
This is necessary for update_authors and update_appdata, for update_statistics it is just for 

To read the variable before setting it is unnecessary (and probably less readable). This just saves a call to git, but git will run fast.

An other option to achive this without a variable is to run update_authors and update_appdata in the beginning. If this changes any files, a 0th commit is created ("updated appdata and authors").
This has the advantage that such changes do not end up either in "translation commit" or "catalog commit" depending on what else changes.
And it probably is more readable.
(Normally, only one or two of the three commits will be triggered.)

Reply:
> The variable run_update_scripts is for case `d)` of [my comment](#issuecomment-1685073340). I want the update_* scripts to run once at least.

OK, then you can remove the `if` in line 171.

> An other option to achive this without a variable is to run update_authors and update_appdata in the beginning. If this changes any files, a 0th commit is created ("updated appdata and authors").

Oh, I see what you mean. Well, for purity's sake we could do that, and they don't change often, so they certainly won't spam the commit history. (and I believe appdata is usually updated right when the stub is changed – mostly on releases)

But still -1 from me, because it's getting unreadable. They can go in the "translation commit", so you don't need to put comments explaining why you run them 3 times.

> This has the advantage that such changes do not end up either in "translation commit" or "catalog commit" depending on what else changes.

I think they can only get updated in the "catalog" commit if there are edits on tx between the 2 pulls, so that's not really an issue.

> And it probably is more readable.

Yes.

Reply:
OK, so if we run them unconditionally below, then this block should be deleted completely.

Reply:
> > An other option to achive this without a variable is to run update_authors and update_appdata in the beginning. If this changes any files, a 0th commit is created ("updated appdata and authors").
> 
> Oh, I see what you mean. Well, for purity's sake we could do that, and they don't change often, so they certainly won't spam the commit history. (and I believe appdata is usually updated right when the stub is changed – mostly on releases)
> 
Right, it was on last release in https://github.com/widelands/widelands/commit/5665009f6b2606ea38fdac58afb3a745a7145848, probably utils/update_appdata.py was run manually. :+1: